src/devices/machine/sis5513_ide.cpp

```diff
@@ -0,0 +1,280 @@
+//	map(0x10, 0x13) Primary channel command base
+//	map(0x14, 0x17) Primary channel control base
+//	map(0x18, 0x1b) Secondary channel command base
+//	map(0x1c, 0x1f) Secondary channel control base
+
+//	map(0x20, 0x23) Bus master IDE control register base (0x10 I/O regs)
+
+//	map(0x2c, 0x2f) subsystem ID (written once)
+
+//	map(0x3c, 0x3d) interrupt line/pin
+
+//	map(0x40, 0x40) IDE Primary channel master data recovery time
+//	map(0x41, 0x41) IDE Primary channel master data active time (Ultra DMA)
+//	map(0x42, 0x42) IDE Primary channel slave data recovery time
+//	map(0x43, 0x43) IDE Primary channel slave data active time (Ultra DMA)
+//	map(0x44, 0x47) ^ Same for IDE Secondary
+
+//	map(0x48, 0x48) IDE status
+//	map(0x4a, 0x4b) IDE general control regs
+
+//	map(0x4c, 0x4d) prefetch count of primary channel
+//	map(0x4e, 0x4f) prefetch count of secondary channel
+
+//	map(0x52, 0x52) IDE misc control regs
+}
+
+void sis5513_ide_device::compatible_io_map(address_map &map)
+{
+	map(0x0170, 0x0177).rw(FUNC(sis5513_ide_device::ide2_read32_cs0_r), FUNC(sis5513_ide_device::ide2_write32_cs0_w));
+	map(0x01f0, 0x01f7).rw(FUNC(sis5513_ide_device::ide1_read32_cs0_r), FUNC(sis5513_ide_device::ide1_write32_cs0_w));
+	map(0x0376, 0x0376).rw(FUNC(sis5513_ide_device::ide2_read_cs1_r), FUNC(sis5513_ide_device::ide2_write_cs1_w));
+	map(0x03f6, 0x03f6).rw(FUNC(sis5513_ide_device::ide1_read_cs1_r), FUNC(sis5513_ide_device::ide1_write_cs1_w));
+}
+
+// $1f0
+void sis5513_ide_device::ide1_command_map(address_map &map)
+{
+	map(0, 7).rw(FUNC(sis5513_ide_device::ide1_read32_cs0_r), FUNC(sis5513_ide_device::ide1_write32_cs0_w));
+}
+
+// $3f4
+void sis5513_ide_device::ide1_control_map(address_map &map)
+{
+	map(2, 2).rw(FUNC(sis5513_ide_device::ide1_read_cs1_r), FUNC(sis5513_ide_device::ide1_write_cs1_w));
+}
+
+// $170
+void sis5513_ide_device::ide2_command_map(address_map &map)
+{
+	map(0, 7).rw(FUNC(sis5513_ide_device::ide2_read32_cs0_r), FUNC(sis5513_ide_device::ide2_write32_cs0_w));
+}
+
+// $374
+void sis5513_ide_device::ide2_control_map(address_map &map)
+{
+	map(2, 2).rw(FUNC(sis5513_ide_device::ide2_read_cs1_r), FUNC(sis5513_ide_device::ide2_write_cs1_w));
+}
+
+void sis5513_ide_device::bus_master_ide_control_map(address_map &map)
+{
+	map(0x0, 0x7).rw(m_ide1, FUNC(bus_master_ide_controller_device::bmdma_r), FUNC(bus_master_ide_controller_device::bmdma_w));
+	map(0x8, 0xf).rw(m_ide2, FUNC(bus_master_ide_controller_device::bmdma_r), FUNC(bus_master_ide_controller_device::bmdma_w));
+}
+
+void sis5513_ide_device::map_extra(uint64_t memory_window_start, uint64_t memory_window_end, uint64_t memory_offset, address_space *memory_space,
+							uint64_t io_window_start, uint64_t io_window_end, uint64_t io_offset, address_space *io_space)
+{
+	bool compatible_mode = BIT(m_ide_misc, 2);
+
+	pclass &= 0xffff80;
+
+	if (compatible_mode)
+	{
+		io_space->install_device(0, 0x03ff, *this, &sis5513_ide_device::compatible_io_map);
```

Comment:
Should this take `io_window_start` into consideration?

Reply:
It's kinda like what is in pci-ide.cpp but not exactly. Compatible mode just overrides effective BARs with legacy addresses but those can be readout with whatever uses sets them up. 
In practice doesn't really belong to map_extra and it's something that never happens if you don't mess with BIOS defaults.

Reply:
ETA: it actually happens with xubuntu 6.10, not closing this up until I have the chance to add an actual implementation.